### PR TITLE
Package Contracts3

### PR DESCRIPTION
**Pull request by mbaluda**
## Description

Contracts3

## Change request type

 - [ ] Release or process automation (GitHub workflows, internal scripts)
 - [ ] Internal documentation
 - [ ] External documentation
 - [x] Query files (`.ql`, `.qll`, `.qls` or unit tests)
 - [ ] External scripts (analysis report or other code shipped as part of a release)

## Rules with added or modified queries

 - [ ] No rules added
 - [x] Queries have been added for the following rules:
     - RULE-22-8 
     - RULE-22-9 
     - RULE-22-10
 - [ ] Queries have been modified for the following rules:
     - _rule number here_

## Release change checklist

A change note ([development_handbook.md#change-notes](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#change-notes)) is required for any pull request which modifies:

 - The structure or layout of the release artifacts.
 - The evaluation performance (memory, execution time) of an existing query.
 - The results of an existing query in any circumstance.

If you are only adding new rule queries, a change note is not required.

_**Author:**_ Is a change note required? 
  - [ ] Yes
  - [x] No

_**Reviewer:**_ Confirm that either a change note is not required or the change note is required and has been added.
  - [x] Confirmed

## Query development review checklist

For PRs that add new queries or modify existing queries, the following checklist should be completed by both the author and reviewer:

### Author

 - [x] Have all the relevant rule package description files been checked in?
 - [x] Have you verified that the metadata properties of each new query is set appropriately?
 - [x] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [x] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [x] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [x] Does the query have an appropriate level of in-query comments/documentation?
 - [x] Have you considered/identified possible edge cases?
 - [x] Does the query not reinvent features in the standard library?
 - [ ] Can the query be simplified further (not golfed!)

### Reviewer 

 - [x] Have all the relevant rule package description files been checked in?
 - [x] Have you verified that the metadata properties of each new query is set appropriately?
 - [x] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [x] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [x] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [x] Does the query have an appropriate level of in-query comments/documentation?
 - [x] Have you considered/identified possible edge cases?
 - [x] Does the query not reinvent features in the standard library?
 - [x] Can the query be simplified further (not golfed!)

### CODE REVIEW

File: c/misra/test/rules/RULE-22-8/test.c [hunk 1]
```diff
@@ -0,0 +1,33 @@
+#include <errno.h>
```
Comment (knewbury01) [inline c/misra/test/rules/RULE-22-8/test.c:1]: do we expect this to be compliant? currently is reported, I think it has something to do with way types are used in Errno.qll lib
```
void f4(void) {
  errno = 0;
  switch(errno) {
    case 0:  
      strtod("0", NULL); // COMPLIANT[FALSE_POSITIVE?]
    default:
      int i =0;
  }
}
```

File: c/common/src/codingstandards/c/Errno.qll [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    this.getLValue() = any(MacroInvocation ma | ma.getMacroName() = "errno").getExpr() and
+    this.getRValue().getValue() = "0"
+  }
+}
+
+/*
+ * A guard controlled by a errno comparison
+ */
+
+class ErrnoGuard extends EqualityOperation {
```
Comment (knewbury01) [inline c/common/src/codingstandards/c/Errno.qll:50]: not sure if its clearer/the same to use a field in this case instead like:
```
class ErrnoGuard extends EqualityOperation {
  ControlStructure i;
  ErrnoGuard() {
    this.getAnOperand() = any(MacroInvocation ma | ma.getMacroName() = "errno").getExpr() and
    this.getAnOperand().getValue() = "0" and
    i.getControllingExpr() = this
  }

  Stmt getThenSuccessor() {
      i.getControllingExpr() = this and
      (result = i.(IfStmt).getThen() or result = i.(Loop).getStmt())
  }

  Stmt getElseSuccessor() {
      i.getControllingExpr() = this and
      (
        i.(IfStmt).hasElse() and result = i.(IfStmt).getElse()
        or
        result = i.getFollowingStmt()
      )
  }
```

File: c/common/src/codingstandards/c/Errno.qll [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+      (
+        i.(IfStmt).hasElse() and result = i.(IfStmt).getElse()
+        or
+        result = i.getFollowingStmt()
+      )
+    )
+  }
+
+  ControlFlowNode getZeroedSuccessor() {
+    if this instanceof EQExpr then result = this.getThenSuccessor() else result = getElseSuccessor()
```
Comment (knewbury01) [inline c/common/src/codingstandards/c/Errno.qll:76]: should one of these be `EQExpr` and the other `NEExpr`? I may be wrong on that

**Review by knewbury01 — COMMENTED** [review]
Cool, good work, few minor suggestions/questions but very neat rules, love the uses of recursion and control flow!

File: rule_packages/c/Contracts3.json [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+        "tags": ["correctness"]
+      }],
+      "title": "The value of errno shall be set to zero prior to a call to an errno-setting-function"
+    },
+    "RULE-22-9": {
+      "properties": {
+        "obligation": "required"
+      },
+      "queries": [{
+        "description": "The value of errno shall be tested against zero after calling an errno-setting-function. Not testing the value leads to undientified errors.",
```
Comment (knewbury01) [inline rule_packages/c/Contracts3.json:38]: "undientified" -> "unidentified"

**Review by knewbury01 — APPROVED** [review]
thanks for addressing all of the feedback!
